src/proc/namer.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{arena::Handle, FastHashMap};
+use std::collections::hash_map::Entry;
+
+pub type EntryPointIndex = u16;
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum NameKey {
+    GlobalVariable(Handle<crate::GlobalVariable>),
+    Type(Handle<crate::Type>),
+    StructMember(Handle<crate::Type>, u32),
+    Function(Handle<crate::Function>),
+    FunctionParameter(Handle<crate::Function>, u32),
+    FunctionLocal(Handle<crate::Function>, Handle<crate::LocalVariable>),
+    EntryPoint(EntryPointIndex),
+    EntryPointLocal(EntryPointIndex, Handle<crate::LocalVariable>),
+}
+
+/// This processor assigns names to all the things in a module
+/// that may need identifiers in a textual backend.
+pub struct Namer {
+    unique: FastHashMap<String, u32>,
+}
+
+const PARAM_NAME: &str = "param";
+
+impl Namer {
+    fn sanitize(string: &str) -> String {
+        let mut base = string
+            .chars()
+            .skip_while(|c| c.is_numeric())
+            .filter(|&c| c.is_ascii_alphanumeric() || c == '_')
+            .collect::<String>();
+        // close the name by '_' if the re is a number, so that
+        // we can have our own number!
+        match base.chars().next_back() {
+            Some(c) if !c.is_numeric() => {}
+            _ => base.push('_'),
+        };
+        base
+    }
+
+    fn call(&mut self, label_raw: &str) -> String {
+        let base = Self::sanitize(label_raw);
+        match self.unique.entry(base) {
+            Entry::Occupied(mut e) => {
+                *e.get_mut() += 1;
+                format!("{}{}", e.key(), e.get())
+            }
+            Entry::Vacant(e) => {
+                let name = e.key().to_string();
+                e.insert(0);
+                name
+            }
+        }
+    }
+
+    fn call_or(&mut self, label: &Option<String>, fallback: &str) -> String {
+        self.call(match *label {
+            Some(ref name) => name,
+            None => fallback,
+        })
+    }
+
+    pub fn process(
+        module: &crate::Module,
+        reserved: &[&str],
+        output: &mut FastHashMap<NameKey, String>,
+    ) {
+        let mut this = Namer {
+            unique: reserved
+                .iter()
+                .map(|string| (string.to_string(), 0))
+                .collect(),
+        };
+        let max_parameters = 100;
+        this.unique.insert(PARAM_NAME.to_string(), max_parameters);
```

Comment:
I am only confused to why we need to insert this here, is this an always reserved keyword?

Reply:
This needs to be reworked...

Reply:
Added a comment to explain